### DISCUSSION TITLE
buggy npm run prod with Amplify Datastore

### DISCUSSION THREAD

**Discussion by charlieforward9** [vikejs/vike#934] [Discussion] [answered]
I am faced today with a worrysome bug. Our application has been under development for about 4 months , and for those 4 months, I have been developing it in my localhost, using `npm run dev`.

I am now at the point where I need to create a test enviornment, which requires HTTPS for the Amplify 3rd party auth Return URL. For HTTPS, cannot serve my application on port 3000, so that leads me to where I am now, running `npm run prod`. 

That gives me this error: 

> Error: ERR_UNSUPPORTED_DIR_IMPORT /Users/crich/Documents/Github/Agroview2.0/node_modules/@aws-amplify/datastore/ssr /Users/crich/Documents/Github/Agroview2.0/dist/server/renderer/default-page-server.js
>     at finalizeResolution (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:362:17)
>     at moduleResolve (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:801:10)
>     at Object.defaultResolve (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:912:11)
>     at /Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/esm.ts:218:35
>     at entrypointFallback (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/esm.ts:168:34)
>     at /Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/esm.ts:217:14
>     at addShortCircuitFlag (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/esm.ts:409:21)
>     at resolve (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/esm.ts:197:12)
>     at resolve (/Users/crich/Documents/Github/Agroview2.0/node_modules/ts-node/src/child/child-loader.ts:15:39)
>     at nextResolve (node:internal/modules/esm/loader:165:28)

I am not too sure how to move forward with this one. This is my `default.page.server.ts`

```typescript
/**
 * @file Server-side rendering of the page
 */

import ReactDOMServer, { renderToString } from "react-dom/server";
import React from "react";
import { PageShell } from "./PageShell";
import { escapeInject, dangerouslySkipEscape } from "vite-plugin-ssr/server";
import icon from "../assets/agroview.ico";
import type { PageContextServer } from "./types";
import { Provider } from "react-redux";
import { store } from "../src/store/store";

// See https://vite-plugin-ssr.com/data-fetching
const passToClient = ["pageProps", "PRELOADED_STATE", "documentProps"];

async function render(pageContext: PageContextServer) {
  const { Page, pageProps } = pageContext;
  let pageHtml;
  // See https://vite-plugin-ssr.com/head
  const { documentProps } = pageContext.exports;
  const title = (documentProps && documentProps.title) || "Agroview";
  const desc =
    (documentProps && documentProps.description) ||
    "The go-to analysis solution for perrenial crop stakeholders";

  if (pageContext.Page) {
    // For SSR pages
    pageHtml = ReactDOMServer.renderToString(
      <PageShell pageContext={pageContext}>
        <Page {...pageProps} />
      </PageShell>
    );
  } else {
    // For SPA pages
    pageHtml = "";
  }

  const documentHtml = escapeInject`<!DOCTYPE html>
    <html lang="en">
      <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="description" content="${desc}" />
        <link rel="icon" type="image/x-icon" href="${icon}" />
        <link href="../renderer/output.css" rel="stylesheet"> 
        <link href="../renderer/styles.css" rel="stylesheet"> 
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/leaflet.draw/1.0.4/leaflet.draw.css"/>
        <script src="https://kit.fontawesome.com/fabd0bd68b.js" crossorigin="anonymous"></script>
        <title>${title}</title>
        </head>
      <body>
        <div id="page-view">${dangerouslySkipEscape(pageHtml)}</div>
      </body>
      <script>
      <!-- Required for aws sdk -->
        if (global === undefined) {
          var global = window;
        }
      </script>
    </html>`;

  return {
    documentHtml,
    pageContext: {
      // We can add some `pageContext` here, which is useful if we want to do page redirection https://vite-plugin-ssr.com/page-redirection
    },
  };
}

async function onBeforeRender(pageContext: PageContextServer) {
  const { Page } = pageContext;
  const pageHtml = renderToString(
    <Provider store={store}>
      <Page />
    </Provider>
  );

  // Grab the initial state from our Redux store
  const PRELOADED_STATE = store.getState();

  return {
    pageContext: {
      PRELOADED_STATE,
      pageHtml,
    },
  };
}

export { render };
export { passToClient };
export { onBeforeRender };
```

and this is the main page of the app, I want it to load exclusively in the client...

```tsx
/**
 * @file Main page of Agroview
 */

import React from "react";
import ClientOnly from "../components/core/ClientOnly";
import Loading from "../components/core/Loading";

function Page() {
  return (
    <div className="h-screen w-screen z-1 absolute">
      <ClientOnly
        load={() => import("../components/map/MapExplorer")}
        fallback={Loading}
      />
    </div>
  );
}

export { Page };
```

I am mostly confused since the main page shouldnt be loading on the server anyways. I would like to continue using this plugin for maximum flexibility (as promoted), but I am aware that Amplify is yet to integrate with this from a SSR point of view.

**Comment by brillout** [comment]
> ERR_UNSUPPORTED_DIR_IMPORT

Have you tried the solution described at https://vite-plugin-ssr.com/common-issues#npm-packages-containing-invalid-code?

**Comment by charlieforward9** [comment]
Here is my `vite.config.ts`
```ts
import react from "@vitejs/plugin-react";
import ssr from "vite-plugin-ssr/plugin";
/** @type {import('vite').UserConfig} */

export default {
  plugins: [react(), ssr()],
  //https://github.com/aws-amplify/amplify-js/issues/9639
  resolve: {
    alias: {
      "./runtimeConfig": "./runtimeConfig.browser", //fix production build
    },
  },
  ssr: {
    // Add npm packages containing invalid code here
    noExternal: ["@aws-amplify"],
  },
};
```

I tried 
aws-amplify
@aws-amplify
@aws-amplify/datastore
@aws-amplify/datastore/ssr

**Reply by brillout** [reply, marked as answer]
The `ERR_UNSUPPORTED_DIR_IMPORT` error should be gone if you add AWS amplify to `ssr.noExternal` (I believe `'@aws-amplify/datastore'` is the one you need to add). Can you confirm?

Once that's fixed, do you get another error?

**Reply by charlieforward9** [reply]
Looks like I mixed up the lib name, `@aws-amplify/datastore` worked.

Upon rerunning, I get this error:

```bash
import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
                      ^^^^^^^^^^^^^^^^
SyntaxError: Named export 'createAsyncThunk' not found. The requested module '@reduxjs/toolkit' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@reduxjs/toolkit';
const { createSlice, createAsyncThunk } = pkg;
```

Would adding '@reduxjs/toolkit' to ssr.noExternal fix this? Or should I take the recommendation in the error message and modify all of my files? A big priority of our application is scalability.

**Reply by brillout** [reply]
https://vite-plugin-ssr.com/broken-npm-package#named-export-not-found

**Reply by charlieforward9** [reply]
Ill have to keep messing around with it, I am getting this error when I try to separate the import statement from the library. Most likely because its not a default import?

**Reply by brillout** [reply]
> Ill have to keep messing around with it

👍 Let me know how that goes.

**Reply by charlieforward9** [reply]
I put the packages into ssr.noExternal since dev speed isnt a major concern. Then I started getting some `ReferenceError: window is not defined` errors. 

Before going any further into the rabbit hole, i decided to check progress on the `npm run dev` build, and got this error:
```
[1] Error: [vite-plugin-ssr@0.4.104][Bug] You stumbled upon a bug in vite-plugin-ssr's source code. Go to https://github.com/brillout/vite-plugin-ssr/issues/new and copy-paste this error (the error alone is usually enough to fix the problem). A maintainer will fix the bug (usually under 24 hours). Don't hesitate to reach out as it makes vite-plugin-ssr more robust.
[1]     at get (/Users/crich/Documents/Github/Agroview2.0/node_modules/vite-plugin-ssr/dist/cjs/utils/require-shim.js:72:41)
[1]     at /node_modules/uuid/index.js:1:23
[1]     at instantiateModule (file:///Users/crich/Documents/Github/Agroview2.0/node_modules/vite/dist/node/chunks/dep-3007b26d.js:52400:15)
```

**Reply by charlieforward9** [reply]
For context, I got the same error about the named export not being found on `uuid`, so I added that to ssr.noExternal as well. Might be the cause of the issue, but I am not using uuid in my source code directly, so not sure how to fix that otherwise.

**Reply by brillout** [reply]
> [vite-plugin-ssr@0.4.104][Bug]

Update to the latest version and let me know if you still get this.

**Reply by charlieforward9** [reply]
Previous error message: 
```bash
import { v4 } from "uuid";
         ^^
SyntaxError: Named export 'v4' not found. The requested module 'uuid' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'uuid';
const { v4 } = pkg;
```

Attempting new VPS package version now

**Reply by charlieforward9** [reply]
The above error persists when:

```diff
-"vite-plugin-ssr": "^0.4.82"
+"vite-plugin-ssr": "^0.4.131"
```

it remains functional when I call `npm run dev`

**Reply by brillout** [reply]
What's the error stack trace with `0.4.131`?

**Reply by charlieforward9** [reply]
```bash
import { v4 } from "uuid";
         ^^
SyntaxError: Named export 'v4' not found. The requested module 'uuid' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'uuid';
const { v4 } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async Object.pageFile.loadFile (/Users/crich/Documents/Github/Agroview2.0/node_modules/vite-plugin-ssr/dist/cjs/shared/getPageFiles/parseGlobResults.js:35:40)
    at async Promise.all (index 1)
    at async loadPageFilesServerSide (/Users/crich/Documents/Github/Agroview2.0/node_modules/vite-plugin-ssr/dist/cjs/shared/getPageFiles/analyzePageServerSide/loadPageFilesServerSide.js:10:5)
    at async Promise.all (index 0)
    at async loadPageFilesServer (/Users/crich/Documents/Github/Agroview2.0/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage/loadPageFilesServer.js:15:110)
    at async renderPageContext (/Users/crich/Documents/Github/Agroview2.0/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage/renderPageContext.js:36:23)
```

**Reply by brillout** [reply]
That's unrelated to VPS. I meant the `[Bug]` one.

**Reply by charlieforward9** [reply]
That is the full error message. I am updating my node from v16 to stable. Hopefully this resolves the uuid issue.

**Reply by brillout** [reply]
I meant this one:

> [1] Error: [vite-plugin-ssr@0.4.104][Bug] You stumbled upon a bug in vite-plugin-ssr's source code. Go to https://github.com/brillout/vite-plugin-ssr/issues/new and copy-paste this error (the error alone is usually enough to fix the problem). A maintainer will fix the bug (usually under 24 hours). Don't hesitate to reach out as it makes vite-plugin-ssr more robust.

**Reply by charlieforward9** [reply]
Yes, I know. Thats the full error message I got.

**Reply by brillout** [reply]
I need the stack trace of:

```diff
- Error: [vite-plugin-ssr@0.4.104][Bug] 
+ Error: [vite-plugin-ssr@0.4.131][Bug] 
```

**Reply by charlieforward9** [reply]
I am unable to reproduce that error in either dev or prod. The `uuid` error is the only thing standing in my way. As I said earlier, I am not using `uuid` directly, so I am not sure why it is giving me an error, especially only in `prod`. `dev` is working fine.

**Reply by brillout** [reply]
VPS doesn't use `uuid`, so I don't think it's related to VPS.

**Comment by charlieforward9** [comment]
If SSR is the issue, I am debating if our application requires SSR since we also currently have a designated marketing site. Is it possible to temporarily disable SSR on the main pages without disintegrating with this package until we have the manpower to create designated SSR pages?

**Reply by brillout** [reply]
Yes, see https://vite-plugin-ssr.com/render-modes.

But keep in mind that removing SSR will make the app's perceived first load noticeably slower. This is usually ok for non-marketing pages but your mileage may vary.

**Reply by charlieforward9** [reply]
Yeah, this is a non-marketing site. There is already an initial page load for the Leaflet map we are using, so hopefully it doesnt add too much more to that. Is there a standard way of creating a server side loading page and passing the SPA page upon load? I figured thats what the ClientOnly component could help with.

**Reply by brillout** [reply]
Yes you should be able to do that.

**Comment by charlieforward9** [comment]
Thank you for all the help. I ended up migrating the project to pure Vite build tooling and the errors went away. This is certainly unfortunate news to you, as I recognize you work very hard on this, but VPS is simply not the right tool for our job.

I want to express my deep appreciation for your extremely rapid feedback. I have never met someone so available on Github, and that definitely sets you apart from other Open Source developers. I will continue to give what I can as a sponsor as an extended thank you for your support, and will try to use this tool for future projects where SSR is more of a priority. 

Thank you, Charles

**Reply by brillout** [reply]
Thanks for the update!

**Reply by brillout** [reply]
As for moving to vanilla Vite: it makes sense if you don't need any of VPS's feature.

FYI, with the upcoming built-in React integration (called `vike-react`), you'll be able to opt-out from SSR simply by setting a config `ssr` to `false`.

The issues you reported in this disucssion are problems with the npm ecosystem (and to some degree Vite). There is, unfortunately, not much vite-plugin-ssr can do about it (I'd definitely fix it otherwise). Although I've an idea to somewhat alleviate the situation (in case you're curious: https://github.com/brillout/vite-plugin-ssr/issues/621#issuecomment-1512789697).

Thanks for sponsoring btw., I appreciate the gesture. (But feel free to cancel, especially now that you don't use VPS anymore.)
